build.rs

```diff
@@ -17,32 +17,117 @@
-
-    } else if target.contains("darwin") {
-        cc::Build::new()
-            .file("etc/hidapi/mac/hid.c")
-            .include("etc/hidapi/hidapi")
-            .compile("libhidapi.a");
-        println!("cargo:rustc-link-lib=framework=IOKit");
-        println!("cargo:rustc-link-lib=framework=CoreFoundation");
-
-    } else if target.contains("linux") {
-        let mut config = cc::Build::new();
-        config.file("etc/hidapi/linux/hid.c").include("etc/hidapi/hidapi");
-        config.compile("libhidapi.a");
-        println!("cargo:rustc-link-lib=udev");
-    }
+	let target = env::var("TARGET").unwrap();
+
+	if target.contains("windows") {
+		cc::Build::new()
+			.file("etc/hidapi/windows/hid.c")
+			.include("etc/hidapi/hidapi")
+			.compile("libhidapi.a");
+		println!("cargo:rustc-link-lib=setupapi");
+
+	} else if target.contains("darwin") {
+		cc::Build::new()
+		.file("etc/hidapi/mac/hid.c")
+		.include("etc/hidapi/hidapi")
+		.compile("libhidapi.a");
+
+		println!("cargo:rustc-link-lib=framework=IOKit");
+		println!("cargo:rustc-link-lib=framework=CoreFoundation");
+
+	} else if target.contains("android") {
+		enable_android_hack();
+		env::set_var("CXX", "arm-linux-androideabi-clang++");
+		env::set_var("CC", "arm-linux-androideabi-gcc");
+		let libudev = autotools::Config::new("etc/eudev")
+			// -s: make symlinks
+			// -m: build if it applicable
+			// -i: install
+			// -v: verbose
+			// -f: consider all files obsolete
+			.reconf("-smivf")
+			.insource(true)
+			.host("arm-linux-androideabi")
+			.disable_shared()
+			.disable("-introspection", None)
+			.disable("-programs", None)
+			.disable("-hwdb", None)
+			.cflag("-D LINE_MAX=2048")
+			.cflag("-D RLIMIT_NLIMITS=15")
+			.cflag("-D IPTOS_LOWCOST=2")
+			.cflag("-std=gnu99")
+			.build();
+
+		disable_android_hack();
+
+		let mut config = cc::Build::new();
+		config.file("etc/hidapi/linux/hid.c").include("etc/hidapi/hidapi");
+		config.compile("libhidapi.a");
+
+		println!("cargo:rustc-link-search=native={}/src/libudev/.libs", libudev.display());
+		println!("cargo:rustc-link-lib=static=udev");
+	} else if target.contains("linux") {
+
+		let libudev = autotools::Config::new("etc/eudev")
+			// -s: make symlinks
+			// -m: build if it applicable
+			// -i: install
+			// -v: verbose
+			// -f: consider all files obsolete
+			.reconf("-smivf")
+			.insource(true)
+			.disable_shared()
+			.build();
+
+		cc::Build::new()
+			.file("etc/hidapi/linux/hid.c")
+			.include("etc/hidapi/hidapi")
+			.compile("libhidapi.a");
+
+		println!("cargo:rustc-link-search=native={}/src/libudev/.libs", libudev.display());
+		println!("cargo:rustc-link-lib=static=udev");
+	}
+}
+
+fn enable_android_hack() {
+	env::set_current_dir(Path::new("etc/eudev"))
+		.expect("Could not find the directory: \"etc\\eudev\"");
+
+	Command::new("git")
+			.args(&["checkout", "83d918449f22720d84a341a05e24b6d109e6d3ae"])
+			.status()
+			.expect("git checkout failed");
+
+	Command::new("git")
```

Comment:
I was planning to use `git2-rs` here but they apparently doesn't support `git_apply_path` AFAIK.

I expect the `cross-compilation` for `android` to fail on Windows but maybe it is possible to configure to be in `Path`